### DISCUSSION TITLE
Don't try to build for too old toolchains

### DISCUSSION THREAD

**Discussion by hassila** [SwiftPackageIndex/SwiftPackageIndex-Server#2294] [Ideas]
Checking for e.g. Benchmark, I get this:
https://swiftpackageindex.com/ordo-one/package-benchmark/builds

The build tries to build for many platforms and toolchain versions that fairly easily could be checked and skipped, either using e.g. by checking:

https://github.com/ordo-one/package-benchmark/blob/main/.swift-version
or
https://github.com/ordo-one/package-benchmark/blob/3b3d46ee240a886d8987938d0c563874e04feeda/Package.swift#L1

one could skip all 5.4, 5.5, 5.6 builds and save some time at your end.

Alternatively you could support expressing this in `.spi.yml`.

**Comment by daveverwer** [comment]
This comes up quite regularly, and we've always decided not to do it. The build itself checks the toolchain and fails almost instantly if it's not compatible, so it doesn't have a huge effect on performance or tying up the build machines. The checking wouldn't introduce much complexity, but it would add some, and it's not really solving a real-world problem.

We'd almost certainly still want to insert "failed" build records for these builds to keep things simple on the display side, so it would introduce another path through to an inserted build record, which would split up that logic a bit.

I vaguely remember this coming up a few months ago, and something in the back of my head says @finestructure had a reason we would consider it, so I'll let him chime in if there's something I've forgotten about the previous discussions.

**Comment by hassila** [comment]
Ok, no worries from my part - just wanted to save you some CPU cycles ;-)

**Reply by finestructure** [reply]
Yeah, it's the dev cycles we're short on though :)

I've been tracking this here: https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/issues/2125

To give you an idea for `package-benchmark`, the outdated build tools builds fail in ~30s, which isn't a problem at all given our current utilisation. Of course, for large repos the checkout we need to do (despite `--depth 1`) can take a little longer, making this a more worthwhile optimisation.

However, as Dave said, we can't just skip the build but would have to short-circuit the trigger directly into a build report and that touches a few places. It's certainly not a huge task but also not a one-liner.

Still waiting for someone eager to want to pick it up as a starter issue 😄
